### PR TITLE
[FLINK-27902][runtime] Refactor ResultPartitionType to decouple scheduling and partition release logic

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

Refactor ResultPartitionType to decouple scheduling and partition release logic

## Brief change log

  - *Introduce ConsumeType indicates when can the downstream consume the upstream. *
  - *Introduce ReleaseType indicates who is responsible for releasing the result partition.*
  - *Using ConsumeType and ReleaseType instead of isPipelined and isReconnectable in scheduling and partition release related code.*
  -  *Remove useless isPipelined, isBlocking and isReconnectable in ResultPartitionType.*

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Review by xintongsong — CHANGES_REQUESTED** [review]
Thanks for opening the PR, @reswqa. The refactor looks good in general. I have only a few comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java [hunk 1]
```diff
@@ -177,6 +177,32 @@ public boolean isReleaseByScheduler() {
...
+     * {@link #isPipelinedOrPipelinedBoundedResultPartition()} is used to judge whether it is the
+     * specified {@link #PIPELINED} or {@link #PIPELINED_BOUNDED} resultPartitionType.
+     *
+     * <p>This method suitable for judgment conditions related to the specific implementation of
+     * {@link ResultPartitionType}.
+     *
+     * <p>This method not related to data consumption and partition release. As for the logic
+     * related to partition release, use {@link #isReleaseByScheduler()} instead, and as consume
+     * type, use {@link #mustBePipelinedConsumed()} or {@link #canBePipelinedConsumed()} instead.
+     */
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java:174]: good catch

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java [hunk 2]
```diff
@@ -104,13 +118,40 @@ public enum ResultPartitionType {
...
+        CAN_BE_PIPELINED,
+        /** Downstream must consume while upstream is running. */
+        MUST_BE_PIPELINED
+    }
+
+    /** ReleaseType indicates who is responsible for releasing the result partition. */
+    private enum ReleaseType {
+        RELEASE_BY_UPSTREAM,
+        RELEASE_BY_SCHEDULER
+    }
```
[the remark this answers is not in the document]
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java]: open [FLINK-27948](https://issues.apache.org/jira/browse/FLINK-27948) to track this issue

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java [hunk 3]
```diff
@@ -104,13 +107,46 @@ public enum ResultPartitionType {
...
+     * ReleaseType indicates who is responsible for releasing the result partition.
+     *
+     * <p>Attention: This may only be a short-term solution to deal with the partition release
+     * logic. We can discuss the issue of unifying the partition release logic in FLINK-27948. Once
+     * the ticket is resolved, we can remove the enumeration here.
+     */
+    private enum ReleaseBy {
+        UPSTREAM,
+        SCHEDULER
+    }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionType.java:116]: Class name in JavaDoc not updated.

**Review by xintongsong — APPROVED** [review]
LGTM. There are only one minor comment, plus that commit messages should indicate the relevant component. I'll fix them while merging this.

**Pull request closed without merging**